chacha20/src/variants.rs

```diff
@@ -2,17 +2,58 @@
 //!
 //! To be revisited for the 64-bit counter.
 
+#[cfg(feature = "cipher")]
+/// A trait to restrict the counter for the cipher crate
+pub trait VariantCounter: cipher::Counter {}
+#[cfg(not(feature = "cipher"))]
+pub trait VariantCounter {}
+
+impl VariantCounter for u32 {}
+
+#[cfg(feature = "legacy")]
+impl VariantCounter for u64 {}
+
 /// A trait that distinguishes some ChaCha variants
 pub trait Variant: Clone {
     /// the size of the Nonce in u32s
     const NONCE_INDEX: usize;
+    /// This const should be evaluated at compile time
+    const USES_U32_COUNTER: bool;
```

Comment:
This seems unnecessary. You already have `Counter = u32`. If nothing else you can use `size_of<Self::Counter>`

Reply:
I've adjusted it to use `size_of`